src/SdkCommon/ClientRuntime/ClientRuntime/ServiceClient.cs

```diff
@@ -384,8 +389,11 @@ protected virtual void Dispose(bool disposing)
                 {
                     HttpClient.Dispose();
                     HttpClient = null;
-                }                
-                
+                }
+
+                if (_disposeRootHandler)
```

Comment:
@EricDahlvang let's just call Dispose on the set the Handlers that are part of ServiceClient. I don't think it's required to have a flag to track it.

Reply:
If the ServiceClient did not create the handler, then it should not dispose it.  This addition only disposes it if it was internally created.

issue ref: https://github.com/Azure/azure-sdk-for-net/issues/3712

Reply:
@EricDahlvang Maybe I am not able to understand your scenario.
Can you tell me a scenario where someone will initialize Service client and would like to use it's RootHandler after disposing the service client.

Reply:
Anyone who uses this constructor:

https://github.com/EricDahlvang/azure-sdk-for-net/blob/e13313279e108e350289dc7373414bcac4e4602a/src/SdkCommon/ClientRuntime/ClientRuntime/ServiceClient.cs#L288

```
        protected ServiceClient(HttpClientHandler rootHandler, params DelegatingHandler[] handlers)
        {
            InitializeHttpClient(rootHandler, handlers);
        }
```

(I don't personally use that constructor, but I saw that this is an issue ... so included it in this PR)

Reply:
It looks like I missed something.  Constructors on lines 250 and 258 also create the RootHandler, and in those cases it should be disposed.

Reply:
@EricDahlvang  is this still in discussion, I found some issue related.I agree with you.It's not thread safe when set the user agent to the http client header.This will lead some critical issue when using http client

Reply:
@mjaow I have no idea if this is still in discussion.  The azure sdk team's previous discussion with me was not fruitful. I'm not actually on the azure sdk team, but was just trying to help.  The team apparently wanted to make the changes themselves.  Things just got worse imo.

- As far as I know, ProductInfoHeaderValues without a product (comment only) will still throw exceptions in this code:
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/mgmtcommon/ClientRuntime/ClientRuntime/ServiceClient.cs#L567

```cs
if (!HttpClient.DefaultRequestHeaders.UserAgent.Contains<ProductInfoHeaderValue>(pInfoHeaderValue,
                    new ObjectComparer<ProductInfoHeaderValue>((left, right) => left.Product.Name.Equals(right.Product.Name, StringComparison.OrdinalIgnoreCase))))
                {
...
```

- Created objects are still not properly disposed.

- The locking strategy in ServiceClient does not make sense to me, and I think ServiceClient should just be a singleton per service.

Reply:
@EricDahlvang , yeah I'm in the same situation with you.I need to create ContainerInstanceManagementClient every request due to the different subscription id and token info.Creating a new ContainerInstanceManagementClient instance will create a new ServiceClient. So the lock seems useless in this condition
I think the right way to fix is to remove the part of setting user agent